staging/src/k8s.io/api/events/v1/types.go

```diff
@@ -0,0 +1,114 @@
+	// +optional
+	Series *EventSeries `json:"series,omitempty" protobuf:"bytes,3,opt,name=series"`
+
+	// Name of the controller that emitted this Event, e.g. `kubernetes.io/kubelet`.
+	// +optional
+	ReportingController string `json:"reportingController,omitempty" protobuf:"bytes,4,opt,name=reportingController"`
+
+	// ID of the controller instance, e.g. `kubelet-xyzf`.
+	// +optional
+	ReportingInstance string `json:"reportingInstance,omitempty" protobuf:"bytes,5,opt,name=reportingInstance"`
+
+	// What action was taken/failed regarding to the regarding object.
+	// +optional
+	Action string `json:"action,omitempty" protobuf:"bytes,6,name=action"`
+
+	// Why the action was taken.
+	Reason string `json:"reason,omitempty" protobuf:"bytes,7,name=reason"`
+
+	// The object this Event is about. In most cases it's an Object reporting controller implements.
+	// E.g. ReplicaSetController implements ReplicaSets and this event is emitted because
+	// it acts on some changes in a ReplicaSet object.
+	// +optional
+	Regarding corev1.ObjectReference `json:"regarding,omitempty" protobuf:"bytes,8,opt,name=regarding"`
+
+	// Optional secondary object for more complex actions. E.g. when regarding object triggers
+	// a creation or deletion of related object.
+	// +optional
+	Related *corev1.ObjectReference `json:"related,omitempty" protobuf:"bytes,9,opt,name=related"`
+
+	// Optional. A human-readable description of the status of this operation.
+	// Maximal length of the note is 1kB, but libraries should be prepared to
+	// handle values up to 64kB.
+	// +optional
+	Note string `json:"note,omitempty" protobuf:"bytes,10,opt,name=note"`
+
+	// Type of this event (Normal, Warning), new types could be added in the
+	// future.
+	// +optional
+	Type string `json:"type,omitempty" protobuf:"bytes,11,opt,name=type"`
+
+	// Deprecated field assuring backward compatibility with core.v1 Event type
+	// +optional
+	DeprecatedSource corev1.EventSource `json:"deprecatedSource,omitempty" protobuf:"bytes,12,opt,name=deprecatedSource"`
+	// Deprecated field assuring backward compatibility with core.v1 Event type
+	// +optional
+	DeprecatedFirstTimestamp metav1.Time `json:"deprecatedFirstTimestamp,omitempty" protobuf:"bytes,13,opt,name=deprecatedFirstTimestamp"`
+	// Deprecated field assuring backward compatibility with core.v1 Event type
+	// +optional
+	DeprecatedLastTimestamp metav1.Time `json:"deprecatedLastTimestamp,omitempty" protobuf:"bytes,14,opt,name=deprecatedLastTimestamp"`
+	// Deprecated field assuring backward compatibility with core.v1 Event type
+	// +optional
+	DeprecatedCount int32 `json:"deprecatedCount,omitempty" protobuf:"varint,15,opt,name=deprecatedCount"`
+}
+
+// EventSeries contain information on series of events, i.e. thing that was/is happening
+// continuously for some time.
+type EventSeries struct {
+	// Number of occurrences in this series up to the last heartbeat time
```

Comment:
this doc references  a heartbeat, but it's unclear which field that refers to... is that deprecatedLastTimestamp or lastObservedTime or something else?

Reply:
It's last UpdateTime.
Basically, the whole EventSeries struct is updated on heartbeat.

Reply:
> It's last UpdateTime.

is there a field on the API object that represents this or is it just implicit?

Reply:
maybe the thing I find confusing is that the heartbeat interval is not visible in the API object, so as an API client looking at an event, I have no idea what the number/lastObservedTime really means if it's not changing...

it could mean the event is not happening any more, or it could mean the event reporter has a really long heartbeat time and the event is happening constantly

is there any guidance or recommendations for reporters on heartbeat time, and for clients on how to interpret these fields?

Reply:
I wasn't that deeply involved in the original design and I can't remember that.

But IIRC, the semantic is really hidden in the client library which is:
- updating it only once per 30 minutes if this is still periodically changing
- closing the series if no event happened in the N minutes (N=6?)

@yastij ^^

Reply:
explaining here that it is up to the event reporter to decide how frequently to update the event, and describing (in general terms) the defaults could help set expectations for consumers and guide non-client-go reporter implementations

Reply:
Sure - that makes perfect sense.

@chelseychen - please update (let me know if you have questions, I can help you framing that)

Reply:
what @wojtek-t described is correct. I agree with @liggitt this might be useful

Reply:
I've updated the doc for EventSeries struct. Not sure if my wording is clear enough. Please let me know what you think, Wojciech. Thanks!